### ISSUE TITLE
CS8625 on call to Interlocked.Exchange with null

### ISSUE THREAD

**Issue by Pilchie** [dotnet/roslyn#31865]
**Version Used**: 
3.0.0-beta2-18616-01

**Steps to Reproduce**:
```C#
using System.Threading;

class C
{
    public C()
    {
        var cts = new CancellationTokenSource();
        Interlocked.Exchange(ref cts, null);
    }
}
``` 

**Expected Behavior**:
No warning

**Actual Behavior**:
Program.cs(8,39,8,43): warning CS8625: Cannot convert null literal to non-nullable reference or unconstrained type parameter.

As Interlocked.Exchange isn't annotated, I can't tell why I'm, getting this warning.

**Comment by Pilchie** [comment]
Similar on:

```C#

using System.Collections.Immutable;

class C<T>
{
    public C()
    {
        var builder = ImmutableArray.CreateBuilder<T>();
        builder.Add(default);
    }
}
```

**Comment by Pilchie** [comment]
Also a `CS8604` for a call to `IImmutableDictionary.SetItem()`

**Comment by cston** [comment]
The warnings are expected.

In the first example, `Interlocked.Exchange(ref cts, null)` is a call to `Interlocked.Exchange<T>(ref T, T) where T : class~`. Since the first parameter is a `ref` parameter, the `CancellationTokenSource!` argument is an exact bounds for `T`. The warning is for the conversion of the second argument `null` to `CancellationTokenSource!`.

In the second example, `builder` is of type `ImmutableArray<T>`, so the conversion of `default(T)` to `T` is a warning. (For instance, if `T` is `string!`, we need to warn on the call to `Add(null)`.)

**Comment by Pilchie** [comment]
Doh!  I'm definitely not used to the behavior around generics yet.
